components/raftstore-v2/src/operation/bucket.rs

```diff
@@ -0,0 +1,202 @@
+    store::{util, Bucket, BucketRange, ReadProgress, SplitCheckTask, Transport},
+};
+use slog::error;
+
+use crate::{batch::StoreContext, fsm::PeerFsmDelegate, raft::Peer, router::PeerTick, worker::pd};
+
+impl<EK: KvEngine, ER: RaftEngine> Peer<EK, ER> {
+    #[inline]
+    pub fn on_refresh_region_buckets<T>(
+        &mut self,
+        store_ctx: &mut StoreContext<EK, ER, T>,
+        region_epoch: RegionEpoch,
+        mut buckets: Vec<Bucket>,
+        bucket_ranges: Option<Vec<BucketRange>>,
+    ) {
+        // bucket version layout
+        //   term       logical counter
+        // |-----------|-----------|
+        //  high bits     low bits
+        // term: given 10s election timeout, the 32 bit means 1362 year running time
+        let gen_bucket_version = |term, current_version| {
+            let current_version_term = current_version >> 32;
+            let bucket_version: u64 = if current_version_term == term {
+                current_version + 1
+            } else {
+                if term > u32::MAX.into() {
+                    error!(
+                        self.logger,
+                        "unexpected term {} more than u32::MAX. Bucket
+                    version will be backward.",
+                        term
+                    );
+                }
+                term << 32
+            };
+            bucket_version
+        };
+
+        let region = self.region();
+        if util::is_epoch_stale(&region_epoch, region.get_region_epoch()) {
+            error!(
+                 self.logger,
+                "receive a stale refresh region bucket message";
+                "epoch" => ?region_epoch,
+                "current_epoch" => ?region.get_region_epoch(),
+            );
+
+            return;
+        }
+        let current_version = self
+            .region_buckets()
+            .as_ref()
+            .or_else(|| self.last_region_buckets().as_ref())
+            .map(|b| b.meta.version)
+            .unwrap_or_default();
+        let mut region_buckets: BucketStat;
+        if let (Some(bucket_ranges), Some(peer_region_buckets)) =
+            (bucket_ranges, self.region_buckets())
+        {
+            assert_eq!(buckets.len(), bucket_ranges.len());
+            let mut i = 0;
+            region_buckets = peer_region_buckets.clone();
+            let mut meta = (*region_buckets.meta).clone();
+            if !buckets.is_empty() {
+                meta.version = gen_bucket_version(self.term(), current_version);
+            }
+            meta.region_epoch = region_epoch;
+            for (bucket, bucket_range) in buckets.into_iter().zip(bucket_ranges) {
+                // the bucket range is the suspected range that needs to split, so it needs to
+                // find the first keys.
+                while i < meta.keys.len() && meta.keys[i] != bucket_range.0 {
+                    i += 1;
+                }
+                assert!(i != meta.keys.len());
```

Comment:
This assertion looks very fragile. What if two `PeerMsg::RefreshRegionBucket` gets scheduled, the first merged some buckets in `self.region_buckets`, and the second doesn't find the key here?

Reply:
It can't happened, the buckets will be reset after [commit_merge](https://github.com/tikv/tikv/blob/7c20add6cef90ac231db6d10374856f23d89e3f1/components/raftstore/src/store/fsm/peer.rs#L4628)

Reply:
I meant bucket merge, not region merge. Buckets in `self.region_buckets.meta` is merged at L106.

Reply:
I will panic if there are two same refreshed message that need to merge the bucket . I will skip this check .

Reply:
![a398fee6-854b-4f27-a0cf-6eab52d0fc2b](https://user-images.githubusercontent.com/23159587/219341299-e8dcfe6f-58de-42af-9191-1328bcf9cb00.jpeg)